### PR TITLE
Use tokenrange to construct lex pres data

### PR DESCRIPTION
**Pull request by matozoid**

### CODE REVIEW

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 1]
```diff
@@ -65,6 +67,7 @@
 
     /**
      * Parse the code and setup the LexicalPreservingPrinter.
+     * @deprecated just use the other constructor.
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:70]: @ftomassetti - I can't see a good reason to have this constructor, is it okay to deprecate?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java [hunk 2]
```diff
@@ -201,7 +205,7 @@ public void process(Node node) {
 
         // We go over tokens and find to which nodes belong. Note that we start from the most specific nodes
         // and we move up to more general nodes
-        for (JavaToken token : documentTokens) {
+        for (JavaToken token : root.getTokenRange().get()) {
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:208]: @ftomassetti Well, having token ranges didn't help much to reduce the code here :-)

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java [hunk 3]
```diff
@@ -21,57 +21,41 @@
...
 
 class TokenTextElement extends TextElement {
-
-    private final int tokenKind;
-    private final String text;
-
-    public static TokenTextElement newLine() {
-        return new TokenTextElement(TokenTypes.eolTokenKind(), EOL);
-    }
+    private final JavaToken token;
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:29]: I'll directly wrap a JavaToken here, so the lex pres code gets a little more connected to the rest.

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java [hunk 4]
```diff
@@ -81,27 +65,25 @@ public boolean equals(Object o) {
...
-        return result;
+        return token.hashCode();
     }
 
     @Override
     public String toString() {
-        return "TokenTextElement(" + tokenKind +
-                ") {" + text + '}';
+        return "TokenTextElement(" + token.getKind() +
+                ") {" + getText() + '}';
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:81]: @ftomassetti I could use JavaToken's toString here - are you very attached to this representation?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:70]: Well, until now it was the easiest way to setup LPP. The user would call this method and get the AST with LPP enabled. How the user should get the same result? Should he parse the code and then call `new LexicalPreservingPrinter(parseResult)`?

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 2 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:208]: Well, now we do not need a ParseResult anymore, just the AST is enough to enable LPP, right?

File: javaparser-core/src/main/java/com/github/javaparser/TokenRange.java [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
+import java.util.Iterator;
 import java.util.Optional;
 
 import static com.github.javaparser.utils.Utils.assertNotNull;
 
 /**
  * The range of tokens covered by this node.
  */
-public class TokenRange {
+public class TokenRange implements Iterable<JavaToken> {
```
Comment (ftomassetti) [inline javaparser-core/src/main/java/com/github/javaparser/TokenRange.java:11]: Nice

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:70]: Oh, ok, with the new changes the Node is enough to start LPP, right

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java (hunk 3 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:29]: ok

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java (hunk 4 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:81]: I am ok with changing it, if it shows the token kind

**Review by ftomassetti — APPROVED** [review]

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:70]: Uhuh, we might even refactor this code into the node constructors ("I know my start token, my end token, and the start and end token of my children, so I can deduce which tokens are mine")

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java (hunk 2 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/LexicalPreservingPrinter.java:208]: Yes, that is an advantage.

File: javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java (hunk 3 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:29]: "Ok, that's great!" or "ok, if you insist..." ?

Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/printer/lexicalpreservation/TokenTextElement.java:29]: The first one: it is great!

**Pull request merged**